### PR TITLE
feat(`cast`): `decode-error` with sig, local cache and openchain api

### PR DESCRIPTION
**Pull request by grandizzy**
## Motivation
- add `cast decode-error [OPTIONS] <DATA>` command: if `--sig` arg specified then it tries to decode data with given signature; if no `--sig` then tries to identify error in local cache or openchain API and decode with. For identified errors, the signature is printed in response
- error sigs can be exported in cache using `forge selectors cache` within project

image

### CODE REVIEW

File: crates/cast/bin/args.rs [hunk 1]
```diff
@@ -539,6 +539,16 @@ pub enum CastSubcommand {
         data: String,
     },
 
+    /// Decode custom error data.
+    #[command(visible_aliases = &["error-decode", "--error-decode", "erd"])]
+    DecodeError {
+        /// The error signature. If none provided then tries to decode from local cache.
+        #[arg(long = "sig", visible_alias = "error-sig")]
```
Comment (zerosnacks) [inline crates/cast/bin/args.rs:546]: `long = "sig"` should not be necessary here I think

File: crates/cast/tests/cli/main.rs [hunk 2]
```diff
@@ -1482,6 +1482,41 @@ casttest!(event_decode, |_prj, cmd| {
 "#]]);
 });
 
+casttest!(error_decode_with_sig, |_prj, cmd| {
+    cmd.args(["decode-error", "--sig", "AnotherValueTooHigh(uint256,address)", "0x7191bc6200000000000000000000000000000000000000000000000000000000000000650000000000000000000000000000000000000000000000000000000000D0004F"]).assert_success().stdout_eq(str![[r#"
+101
+0x0000000000000000000000000000000000D0004F
+
+"#]]);
+});
```
Comment (zerosnacks) [inline crates/cast/tests/cli/main.rs:1491]: Would you mind adding an additional case for the `--json` output here?

File: crates/cast/bin/args.rs (hunk 1 above)
Comment (grandizzy) [reply crates/cast/bin/args.rs:546]: yeah, I want to have sig non required but able to be specified before data, if not adding long it fails with `Found non-required positional argument with a lower index than a required positional argument: "sig" index Some(1)` is there a different way than workaround it by `long = "sig"`?

Comment (grandizzy) [reply crates/cast/bin/args.rs:546]: ah, nm, I will only add `long`

Comment (grandizzy) [reply crates/cast/bin/args.rs]: added in https://github.com/foundry-rs/foundry/pull/9428/commits/d349c9a3cbf22e65db5670ed3d047712926642ea

File: crates/cast/tests/cli/main.rs (hunk 2 above)
Comment (grandizzy) [reply crates/cast/tests/cli/main.rs]: added in https://github.com/foundry-rs/foundry/pull/9428/commits/d349c9a3cbf22e65db5670ed3d047712926642ea

**Review by zerosnacks — APPROVED** [review]
lgtm! :100:

File: crates/evm/traces/src/identifier/signatures.rs [hunk 3]
```diff
@@ -98,32 +99,36 @@ impl SignaturesIdentifier {
         identifiers: impl IntoIterator<Item = impl AsRef<[u8]>>,
         get_type: impl Fn(&str) -> eyre::Result<T>,
     ) -> Vec<Option<T>> {
-        let cache = match selector_type {
-            SelectorType::Function => &mut self.cached.functions,
-            SelectorType::Event => &mut self.cached.events,
+        let (cache, with_openchain) = match selector_type {
+            SelectorType::Function => (&mut self.cached.functions, true),
+            SelectorType::Event => (&mut self.cached.events, true),
+            // Openchain API does not support custom errors.
```
Comment (DaniPopes) [inline crates/evm/traces/src/identifier/signatures.rs:105]: It does, it's in the same format as functions

**Review by DaniPopes — CHANGES_REQUESTED** [review]

File: crates/evm/traces/src/identifier/signatures.rs (hunk 3 above)
Comment (grandizzy) [reply crates/evm/traces/src/identifier/signatures.rs]: oh, was looking at their API and seen fns and events so I assumed errors not supported, thanks for flagging. Added support and a test with https://github.com/foundry-rs/foundry/pull/9428/commits/35f5dc101c3450773423d54c43be0c24a6817d4e please check

File: crates/cast/bin/main.rs [hunk 4]
```diff
@@ -216,6 +217,28 @@ async fn main_args(args: CastArgs) -> Result<()> {
...
+                get_error(err_sig.as_str())?
+            } else {
+                let data = data.strip_prefix("0x").unwrap_or(data.as_str());
+                let selector = &data[..8];
+                let err = SignaturesIdentifier::new(Config::foundry_cache_dir(), false)?
+                    .write()
+                    .await
+                    .identify_error(&hex::decode(selector)?)
+                    .await;
+                if err.is_none() {
```
Comment (DaniPopes) [inline crates/cast/bin/main.rs:231]: let Some ... else

File: crates/cast/bin/args.rs [hunk 5]
```diff
@@ -539,6 +539,16 @@ pub enum CastSubcommand {
         data: String,
     },
 
+    /// Decode custom error data.
+    #[command(visible_aliases = &["error-decode", "--error-decode", "erd"])]
+    DecodeError {
+        /// The error signature. If none provided then tries to decode from local cache or `https://api.openchain.xyz`.
+        #[arg(long, visible_alias = "error-sig")]
+        sig: Option<String>,
```
Comment (DaniPopes) [inline crates/cast/bin/args.rs:547]: can we make this optional for decode event too?

**Review by DaniPopes — COMMENTED** [review]
lgtm nits

File: crates/cast/bin/main.rs (hunk 4 above)
Comment (grandizzy) [reply crates/cast/bin/main.rs]: changed in https://github.com/foundry-rs/foundry/pull/9428/commits/1834215d256f0afb1d574ae538c03b67d6ed5eb1

File: crates/cast/bin/args.rs (hunk 5 above)
Comment (grandizzy) [reply crates/cast/bin/args.rs:547]: yep, will follow up with a PR, have to figure out how to parse with topics

**Review by DaniPopes — APPROVED** [review]

**Pull request merged**